### DISCUSSION TITLE
StackOverflowError adding alot of vertex properties inside one traverser

### DISCUSSION THREAD

**Discussion by mrckzgl** [JanusGraph/janusgraph#3492] [Q&A]
- Version: 0.6.1
- Storage Backend: berkleyje
- Mixed Index Backend: lucene
- Expected Behavior: No Stackoverflow Error
- Current Behavior: Stackoverflow Error

Running this gremlin query inside JRuby:
```
traversal = dbtx.traversal.add_v()

large_list.each{|v|
  traverser.property(VertexProperty::Cardinality::list, 'search_value', v["value"])
}

traverser.iterate()
```
results in:
```
2023-01-20 16:15:48 - Java::JavaLang::StackOverflowError - :
	org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(org/apache/tinkerpop/gremlin/process/traversal/step/util/ExpandableStepIterator.java:55)
	org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.SideEffectStep.processNextStart(org/apache/tinkerpop/gremlin/process/traversal/step/sideEffect/SideEffectStep.java:38)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(org/apache/tinkerpop/gremlin/process/traversal/step/util/AbstractStep.java:150)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(org/apache/tinkerpop/gremlin/process/traversal/step/util/ExpandableStepIterator.java:55)
	org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.SideEffectStep.processNextStart(org/apache/tinkerpop/gremlin/process/traversal/step/sideEffect/SideEffectStep.java:38)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(org/apache/tinkerpop/gremlin/process/traversal/step/util/AbstractStep.java:150)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(org/apache/tinkerpop/gremlin/process/traversal/step/util/ExpandableStepIterator.java:55)
	org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.SideEffectStep.processNextStart(org/apache/tinkerpop/gremlin/process/traversal/step/sideEffect/SideEffectStep.java:38)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(org/apache/tinkerpop/gremlin/process/traversal/step/util/AbstractStep.java:150)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(org/apache/tinkerpop/gremlin/process/traversal/step/util/ExpandableStepIterator.java:55)
	org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.SideEffectStep.processNextStart(org/apache/tinkerpop/gremlin/process/traversal/step/sideEffect/SideEffectStep.java:38)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(org/apache/tinkerpop/gremlin/process/traversal/step/util/AbstractStep.java:150)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.ExpandableStepIterator.next(org/apache/tinkerpop/gremlin/process/traversal/step/util/ExpandableStepIterator.java:55)
	org.apache.tinkerpop.gremlin.process.traversal.step.sideEffect.SideEffectStep.processNextStart(org/apache/tinkerpop/gremlin/process/traversal/step/sideEffect/SideEffectStep.java:38)
	org.apache.tinkerpop.gremlin.process.traversal.step.util.AbstractStep.hasNext(org/apache/tinkerpop/gremlin/process/traversal/step/util/AbstractStep.java:150)
```

For a list of size ~4000.

Not sure if this is a gremlin issue or a JanusGraph issue. Anyway the apache tinkerpop github repo does not have an issue tracker, so I would be very grateful if this is a gremlin issue to get advise where to post the issue.

best

EDIT: Sorry wrong point in source code

**Comment by mrckzgl** [comment]
ok, as I opened the issue I was not aware that we are actually doing ~4000 `traverser.property` calls. Refactoring that to pass a large list to one property call instead of one value to a lot of property calls:
```
traverser.property(VertexProperty::Cardinality::list, 'search_value', large_list.map{|p|p["value"]})
```
does not generate the stack overflow luckily. Still, I would be very interested to hear opinions if the OP is considered a bug or a legitimate limitation (which IMHO should still be documented in the gremlin reference)?

**Comment by mrckzgl** [comment]
ok I just found out in the tinkerpop reference, that it is *not* possible to provide an array / list of values in a `.property` step, which is very weird honestly and unexpected, as properties with cardinality `list` explicitly exists. What actually happens when providing a list (but only if the property key already exists) is that for this list .toString() is called and the string representation of that list is added as value. This implementation will lead to very hard to debug problems. Instead of casting to string an error should be thrown IMHO if an unsupported value type is provided by `.property`.

Anyhow, this also means that above workaround is not possible and the current implementation definitively is a bug IMHO.

Another workaround might be to split the addition of propery values into multiple traversers, sth like this:
```
node = dbtx.traversal.add_v().next

traverser = dbtx.traversal.v(node)
first_list_of_x_values.each{|v|
  traverser.property(VertexProperty::Cardinality::list, 'search_value', v["value"])
}
traverser.iterate()

traverser = dbtx.traversal.v(node)
second_list_of_x_values.each{|v|
  traverser.property(VertexProperty::Cardinality::list, 'search_value', v["value"])
}
traverser.iterate()

...
```
Haven't tested if it works, anyhow it is quite unconvenient, and also impossible to know the exact value of x ...

**Comment by porunov** [comment]
TinkerPop uses Jira as an issue tracker. You can post your TinkerPop found issues here: https://issues.apache.org/jira/projects/TINKERPOP/issues/.
That said, before posting an issue it's usually good to discuss it first.
Discord to ask TinkerPop / Gremlin related questions: https://discord.com/invite/kfebyvDM9H
Gremlin-users Google group to ask TinkerPop / Gremlin related questions: https://groups.google.com/g/gremlin-users
